Introduce `--threads=max` option value to automatically detect the number of CPU cores

maks-rafalko opened the pull request:
Implements https://github.com/infection/infection/issues/1722

Doc PR: https://github.com/infection/site/pull/240

sanmai commented on src/Resource/Processor/CpuCoresCountProvider.php:
```diff
@@ -0,0 +1,110 @@
+        if (ini_get('pcre.jit') === '1'
+            && PHP_OS === 'Darwin'
+            && version_compare(PHP_VERSION, '7.3.0') >= 0
+            && version_compare(PHP_VERSION, '7.4.0') < 0
```
Should we even care about PHP 7.4?

sanmai commented on src/Resource/Processor/CpuCoresCountProvider.php:
```diff
@@ -0,0 +1,110 @@
+            }
+        }
+
+        throw new LogicException('Failed to detect number of CPUs!');
```
Why not `return 1;` in this case?

sanmai commented on src/Resource/Processor/CpuCoresCountProvider.php:
```diff
@@ -0,0 +1,110 @@
+            }
+        }
+
+        $ncpu = trim(shell_exec('sysctl -n hw.ncpu'));
```
It is unclear why we assume `nproc` may not be there while `sysctl` supposed to be always there. Should we even check that `nproc` exists? Because if it isn't clearly we won't be getting a single digit :thinking:

maks-rafalko replied:
this is for MacOS case, where `nproc` is not available so the code will be here, and `sysctl -n hw.ncpu` will return the number of cores. Checked on Linus and Macos machines, added comments

sanmai reviewed (changes requested):
Left few comments, mostly nits. The only real deal is `LogicException` at the end of `provide()`.

sidz commented:
My 2 cents: it is not Infection responsibility. We just make our codebase more complicated. `-j$(nproc)` or `-j$(sysctl -n hw.ncpu)` works fine. End user can use Make or composer scripts if he don't want to type `--threads=`

sidz commented on src/Resource/Processor/CpuCoresCountProvider.php:
```diff
@@ -0,0 +1,110 @@
+
+        if (ini_get('pcre.jit') === '1'
+            && PHP_OS === 'Darwin'
+            && version_compare(PHP_VERSION, '7.3.0') >= 0
```
can't we use `PHP_VERSION_ID` which contains an integer value instead of trying to compare strings?

maks-rafalko commented:
Addressed your comments, initial version was blindly copied from Psalm.

> My 2 cents: it is not Infection responsibility. We just make our codebase more complicated.

Well, yes and no. IMO it's all about "those tiny things for DX". You set `--threads=max` and 

- don't care about any potential cases where `$(nproc)` might not work because of incorrect escaping (in `Jenkinsfile` you will have to prepend it like `sh 'infection ... --threads=\$(nproc)`'`
- don't care and just let anyone copy this command line from Linux to MacOS machines and vice-versa

Yes it's completely not critical and both points can be ignored, but it's a little thing that might be more convenient for someone (which is why this feature request was born from our end user, not me :) ).

sanmai reviewed (approved):
I agree this is a very dangerous option, yet if people want this I guess that's what we have to add :thinking:

dkarlovi commented:
@sanmai what makes it dangerous?

sanmai commented:
- Infection does surprising things with the code. For one, it creates infinite loops. And could include something consuming memory. It could make you machine unresponsive for prolonged periods of time and generally lead to bad experiences, especially if that's the first time you're using Infection.
- Running Infection with more threads not necessary leads to better performance. Unlike, say, pure CPU-bound tasks such as static analysis. Thus this option could give a false promise of fast execution where in fact it will work in the opposite way.

I would seriously object enabling this option by default, or recommending it to new users.

dkarlovi commented:
That means the threads upper limit should be restricted / validated regardless of how you specify it, which further proves the idea Infection should know the upper limit.

For example, if max=12, what difference is it if I say --threads=12 or --threads=max, from the user POV this is totally equivalent. If running it like that is "dangerous", infection should disallow or warn the user about it.

Adding --threads=max seems unrelated (with using max threads being "dangerous") to me.

sanmai commented:
Infection can't know the upper limit. Say, if your test suite is predominantly IO-bound, then it makes sense to run Infection with more threads than you have CPU. 

Do you think there's a sure way what to do in this situation? How can we know if a test suite is CPU-bound or IO-bound to get the best results?

dkarlovi commented:
This is knowledge you have, the user doesn't. It sounds like Infection could have named threads calculator strategies which the user can try out and use some which make sense.

For example, `conservative`, `per-cpu`, `io-bound`, etc.

dkarlovi commented:
@maks-rafalko what do you think here?

sanmai commented:
> threads calculator strategies

How we should approach this idea, given Infection is a PHP app?

dkarlovi commented:
@sanmai not sure what you mean here?

Since now infection knows the number of CPUs available, I can imagine you could do something like
- `--threads=conservative` = number of CPUs * 0.8
- `--threads=per-cpu` = number of CPUs, which is current max
- `--threads=io-bound` = number of CPUs * 1.5

All names and numbers are illustrational, of course.

theofidry commented on src/Resource/Processor/CpuCoresCountProvider.php:
```diff
@@ -0,0 +1,99 @@
+/**
+ * @internal
+ */
+final class CpuCoresCountProvider
```
What do you think of `CpuInfo::getCount()` as a name instead?

theofidry commented on src/Resource/Processor/CpuCoresCountProvider.php:
```diff
@@ -0,0 +1,99 @@
+     */
+    public static function provide(): int
+    {
+        if (defined('PHP_WINDOWS_VERSION_MAJOR')) {
```
Might be worth caching the result? Although I don't expect it to be called twice, you also don't expect the system to change the number of CPU cores available within the same process and it's only a `int` value

theofidry commented on src/Resource/Processor/CpuCoresCountProvider.php:
```diff
@@ -0,0 +1,99 @@
+final class CpuCoresCountProvider
+{
+    /**
+     * Copied and adapter from Psalm project: https://github.com/vimeo/psalm/blob/4.x/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php#L1454
```
Is there any chance Psalm is willing to extract it to a dedicated package? Otherwise provided they would be interested in a package, I'd be happy to provide one

dkarlovi replied:
This might be interesting as some sort of tiny "system information" HAL package which could be reused across the ecosystem indeed. :+1:

theofidry commented:
Sorry for the late review, great addition! Happy to do the changes myself
